flask/app/routes.py

```diff
@@ -244,6 +244,51 @@ def bulk_update():
         updated_by_id = 1
         created_by_id = 1
 
+    # get all group codes present in db
+    all_groups = models.Group.query.all()
+    all_codes = [g.group_code for g in all_groups]
+
+    # get user's group(s)
+    group_query = request.args.get("group")
+
+    if app.config.get("LOGIN_DISABLED"):
+        user_id = request.args.get("user")
+        if user_id:
+            user_group = (
+                models.Group.query.join(models.Group.users)
+                .filter(models.User.user_id == user_id)
+                .all()
+            )
+        else:
+            user_group = all_groups
+    elif current_user.is_admin:
+        user_group = all_groups
+    else:
+        user_id = current_user.user_id
+        user_group = (
+            models.Group.query.join(models.Group.users)
+            .filter(models.User.user_id == user_id)
+            .all()
+        )
+
+    if not user_group:
+        return "User does not belong to any permission groups", 403
+
+    if group_query:
+        group_query = group_query.split(",")
+        codes = [code.group_code for code in user_group]
+        for g in group_query:
+            if g not in all_codes:
+                return f"Group {g} does not exist", 404
+            if g not in codes:
+                return f"User does not belong to group {g}", 403
```

Comment:
https://github.com/ccmbioinfo/stager/blob/master/flask/app/users.py#L182-L189

Can you adopt this approach to checking if the groups exist and the user is allowed access? This makes the database do the work. You'll need to adapt by doing your join and filter above but this should otherwise simply this a lot.

Reply:
simplified in 60361f91d6c4c01f7d8a8fbd256ba82b058a0164